packages/compiler/core/types.ts

```diff
@@ -414,163 +409,263 @@ export type Expression =
+  readonly properties: (ModelPropertyNode | ModelSpreadPropertyNode)[];
 }
 
 export interface ArrayExpressionNode extends BaseNode {
-  kind: SyntaxKind.ArrayExpression;
-  elementType: Expression;
+  readonly kind: SyntaxKind.ArrayExpression;
+  readonly elementType: Expression;
 }
 export interface TupleExpressionNode extends BaseNode {
-  kind: SyntaxKind.TupleExpression;
-  values: Expression[];
+  readonly kind: SyntaxKind.TupleExpression;
+  readonly values: readonly Expression[];
 }
 
 export interface ModelPropertyNode extends BaseNode {
-  kind: SyntaxKind.ModelProperty;
-  id: IdentifierNode | StringLiteralNode;
-  value: Expression;
-  decorators: DecoratorExpressionNode[];
-  optional: boolean;
-  default?: Expression;
+  readonly kind: SyntaxKind.ModelProperty;
+  readonly id: IdentifierNode | StringLiteralNode;
+  readonly value: Expression;
+  readonly decorators: readonly DecoratorExpressionNode[];
+  readonly optional: boolean;
+  readonly default?: Expression;
 }
 
 export interface ModelSpreadPropertyNode extends BaseNode {
-  kind: SyntaxKind.ModelSpreadProperty;
-  target: TypeReferenceNode;
+  readonly kind: SyntaxKind.ModelSpreadProperty;
+  readonly target: TypeReferenceNode;
 }
 
 export type LiteralNode = StringLiteralNode | NumericLiteralNode | BooleanLiteralNode;
 
 export interface StringLiteralNode extends BaseNode {
-  kind: SyntaxKind.StringLiteral;
-  value: string;
+  readonly kind: SyntaxKind.StringLiteral;
+  readonly value: string;
 }
 
 export interface NumericLiteralNode extends BaseNode {
-  kind: SyntaxKind.NumericLiteral;
-  value: number;
+  readonly kind: SyntaxKind.NumericLiteral;
+  readonly value: number;
 }
 
 export interface BooleanLiteralNode extends BaseNode {
-  kind: SyntaxKind.BooleanLiteral;
-  value: boolean;
+  readonly kind: SyntaxKind.BooleanLiteral;
+  readonly value: boolean;
 }
 
 export interface UnionExpressionNode extends BaseNode {
-  kind: SyntaxKind.UnionExpression;
-  options: Expression[];
+  readonly kind: SyntaxKind.UnionExpression;
+  readonly options: readonly Expression[];
 }
 
 export interface IntersectionExpressionNode extends BaseNode {
-  kind: SyntaxKind.IntersectionExpression;
-  options: Expression[];
+  readonly kind: SyntaxKind.IntersectionExpression;
+  readonly options: readonly Expression[];
 }
 
 export interface TypeReferenceNode extends BaseNode {
-  kind: SyntaxKind.TypeReference;
-  target: MemberExpressionNode | IdentifierNode;
-  arguments: Expression[];
+  readonly kind: SyntaxKind.TypeReference;
+  readonly target: MemberExpressionNode | IdentifierNode;
+  readonly arguments: readonly Expression[];
 }
 
 export interface TemplateParameterDeclarationNode extends BaseNode {
-  kind: SyntaxKind.TemplateParameterDeclaration;
-  id: IdentifierNode;
+  readonly kind: SyntaxKind.TemplateParameterDeclaration;
+  readonly id: IdentifierNode;
   symbol?: TypeSymbol;
 }
 
+//#region BoundSyntaxTree
+// prettier-ignore
+export type BindNode<T extends Node> = 
```

Comment:
The tense here feels a little strange to me since it's not doing any binding per se - maybe `BoundNode` is more appropriate?

Reply:
Also, how/where is this used? Not finding any references but maybe I'm missing it.

Reply:
it isn't used anymore, added it orgiinally as I tried changing `bindNode` to return the bindedNode instead of modifying in place. `bindNode(node: T) -> BindNode<T>`.
But as talked about this is doing a bit of perf hit relative to the parser and it was causing other issues as well(bigger change than expected)